### DISCUSSION TITLE
Should `atomWithStorage` in `v2.0.2` work with experimental app dir in Next JS 13?

### DISCUSSION THREAD

**Discussion by CombeeMike** [pmndrs/jotai#1798] [Q&A]
When using `atomWithStorage` in NextJS 13 with the experimental app dir enabled, I'm seeing many errors indicating that hydration did not work. The following 2 are just examples of the errors I'm seeing:
* > ... Uncaught Error: Hydration failed because the initial UI does not match what was rendered on the server. ...
* > ... Warning: Prop `className` did not match. Server: "grid justify-items-center gap-2 grid-cols-fill-icon-size-s" Client: "grid justify-items-center gap-2 grid-cols-fill-icon-size-l" ... (`@ hydration-error-info.js`)
  
  I'm setting different tailwind classes based on the value of an atom using something like `classNames('...', { 'optional-class': valueFromAtom === 'xyz' })`.

Here's how I'm using `atomWithStorage`:

**app.store.ts**
```ts
export const IconSizes = ['S', 'M', 'L', 'XL'] as const;
export type IconSize = (typeof IconSizes)[number];
export const galleryIconSizeAtom = atomWithStorage<IconSize>('icon-size-key', 'S');

export const galleryShowCategoriesAtom = atomWithStorage<boolean>('show-categories-key', true);
```

I'm only importing the atoms from that file within client component (i.e. cmps starting with the `use client;` directive) and am using them solely with `useAtomValue` ATM.

I've seen [this issue](https://github.com/pmndrs/jotai/issues/1689) but don't think that the suggested solution with the `delayInit: true` option applies to Jotai `v2.0.2` anymore.

Is this supposed to be working this way with NextJS 13s beta features, is this a bug or am I doing something wrong here?

**P.s.:** I've not created an interactive reproduction playground yet as I first wanted to clarify whether there's something fundamentally wrong with my approach or not as I just started using Jotai and might be missing some basic understandings.

**Comment by dai-shi** [comment]
> the suggested solution with the `delayInit: true` option applies to Jotai `v2.0.2` anymore.

In v2, `delayInit` is always true.

Anyway, I'm not sure if appDir works in the first place. Does it work without `atomWithStorage`?

**Comment by CombeeMike** [comment]
I did not experience any issues with "normal" `atom`s yet, but that doesn't necessarily mean too much as I just started using Jotai yesterday in this project and have not tested too much yet.

**Comment by lveillard** [comment]
In our case, not sure why, but our atomWithStorage works properly in nextjs 14 with pageDir, but when migrating to the app Dir it causes an infinite re-render.

Not caused by anything external, is like if the atomWithStorage was forcing the atoms that depend on it infinite times.

This happens to us even with:
- A custom store in the layout
- A OnlyClient wrapper
